include/camp/resource/sycl.hpp

```diff
@@ -44,94 +44,171 @@ namespace resources
-                sycl::queue(*contextInUse, gpuSelector, propertyList),
-                sycl::queue(*contextInUse, gpuSelector, propertyList),
-                sycl::queue(*contextInUse, gpuSelector, propertyList)};
+        } else {
+          if (prevContextIter != queueMap_end) {
+            // intentionally left empty as syclContext not needed
+          } else {
+            syclContext = &get_thread_default_context();
           }
         }
-        m_mtx.unlock();
 
-        static int previous = 0;
+        if (prevContextIter == queueMap_end || num < 0) {
+          std::lock_guard<std::mutex> lock(s_mtx);
+
+          if (prevContextIter == queueMap_end) {
+            prevContextIter = queueMap.find(syclContext);
+            if (prevContextIter == queueMap_end) {
+              static constexpr auto gpuSelector = sycl::gpu_selector_v;
+              static const sycl::property_list propertyList =
+                  sycl::property_list(sycl::property::queue::in_order());
+
+              prevContextIter = queueMap.emplace(syclContext, { num_queues-1, {
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList),
+                  sycl::queue(*syclContext, gpuSelector, propertyList)}}).first;
+            }
+          }
 
-        static std::once_flag m_onceFlag;
-        CAMP_ALLOW_UNUSED_LOCAL(m_onceFlag);
-        if (num < 0) {
-          m_mtx.lock();
-          previous = (previous + 1) % 16;
-          m_mtx.unlock();
-          return &queueMap[contextInUse][previous];
+          if (num < 0) {
+            int& previous = prevContextIter->second.first;
+            previous = (previous + 1) % num_queues;
+            return &prevContextIter->second.second[previous];
+          }
         }
 
-        return &queueMap[contextInUse][num % 16];
+        return &prevContextIter->second.second[num % num_queues];
       }
 
       // Private from-queue constructor
       Sycl(sycl::queue& q) : qu(&q) {}
```

Comment:
I see that the Sycl resource holds pointers to queues but how about we hold copies instead? It looks like contexts, queues, and various other sycl classes have reference (shared_ptr) semantics, see https://github.khronos.org/SYCL_Reference/iface/common-reference.html. Holding pointers to queues makes the Sycl resource point to memory owned by other code which is not the case for other resources and easy to mess up.

Reply:
If we switch to copies we should probably also switch `Sycl::get_queue()` to return references instead of pointers. I think we can make this change and the removal of the implicit sticky behavior in a latter pull request.